File: partiql-plan/src/main/java/org/partiql/plan/rex/RexSelect.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rel.Rel;
import org.partiql.types.PType;

import java.util.List;

/**
 * Logical select expression abstract base class.
 */
public abstract class RexSelect extends RexBase {

    /**
     * @return new RexSelect instance
     */
    @NotNull
    public static RexSelect create(@NotNull Rel input, @NotNull Rex constructor) {
        return new Impl(input, constructor);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    /**
     * @return constructor rex.
     */
    public abstract Rex getConstructor();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.bag());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getInput());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitSelect(this, ctx);
    }

    private static class Impl extends RexSelect {

        private final Rel input;
        private final Rex constructor;

        private Impl(Rel input, Rex constructor) {
            this.input = input;
            this.constructor = constructor;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public Rex getConstructor() {
            return constructor;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexSelect.kt
package org.partiql.plan.rex

import org.partiql.plan.rel.Rel
import org.partiql.types.PType

/**
 * TODO DOCUMENTATION
 */
public interface RexSelect : Rex {

    public fun getInput(): Rel

    public fun getConstructor(): Rex

    override fun getChildren(): Collection<Rex> = listOf(getConstructor())

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitSelect(this, ctx)
}

internal class RexSelectImpl(input: Rel, constructor: Rex) : RexSelect {

    private var _input = input
    private var _constructor = constructor
    private var _type: RexType? = null

    override fun getInput(): Rel = _input

    override fun getConstructor(): Rex = _constructor

    override fun getType(): RexType {
        // compute type
        if (_type == null) {
            val e = _constructor.getType().getPType()
            _type = if (_input.isOrdered()) {
                RexType(PType.array(e))
            } else {
                RexType(PType.bag(e))
            }
        }
        return _type!!
    }

    override fun getChildren(): Collection<Rex> = listOf(_constructor)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexSelect) return false
        if (_input != other.getInput()) return false
        if (_constructor != other.getConstructor()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = _input.hashCode()
        result = 31 * result + _constructor.hashCode()
        return result
    }
}
